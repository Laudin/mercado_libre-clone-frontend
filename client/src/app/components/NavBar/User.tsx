import * as React from 'react'
import styled from 'styled-components/macro'

export function User() {
   return (
      <Wrapper>
         <Item><Link>Crear Cuenta</Link></Item>
         <Item><Link>Ingresar</Link></Item>
         <Item><Link>Mis Compras</Link></Item>
         <Item><Link>Carrito</Link></Item>
      </Wrapper>
   )
}

const Wrapper = styled.ul`
   display: flex;
   gap: 10px;
   list-style: none;
`
const Item = styled.li`
`
const Link = styled.a`
   color: black;
   text-decoration: none;
   cursor: pointer;
`
