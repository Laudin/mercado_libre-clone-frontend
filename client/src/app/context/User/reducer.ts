import { User } from '../../../types'

export const reducer = (state, action) => {
  switch (action.type) {
    case "set_user":
      localStorage.setItem('id', action.id)
      localStorage.setItem('name', action.name)
      localStorage.setItem('email', action.email)
      return {
        ...state,
        id: action.id,
        name: action.name,
        email: action.email
      }
    case "clear_user":
      localStorage.removeItem('id')
      localStorage.removeItem('name')
      localStorage.removeItem('email')
      return {
        ...state,
        id: '',
        name: '',
        email: ''
      }

    default:
      return state
  }
}

export const initialState: User = {
  id: localStorage.getItem('id') as string,
  name: localStorage.getItem('name') as string,
  email: localStorage.getItem('email') as string
}