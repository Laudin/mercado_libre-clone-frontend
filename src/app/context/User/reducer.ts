import { User } from '../../../types'

export const reducer = (state, action?) => {
  console.log(action)
  switch (action.type) {
    case "set_user":
      //localStorage.setItem('id', action.id)
      localStorage.setItem('name', action.name)
      localStorage.setItem('email', action.email)

      document.cookie = `id=${action.id}; max-age=${30 * 24 * 60 * 60}; Secure` //sets User Id 
      return {
        ...state,
        id: action.id,
        name: action.name,
        email: action.email
      }
    case "clear_user":
      //localStorage.removeItem('id')
      localStorage.removeItem('name')
      localStorage.removeItem('email')

      document.cookie = `id=${action.id}; expires=Thu, 01 Jan 1970 00:00:01 GMT"; max-age=${0}; Secure` // "kills" the cookie x_x
      return {
        ...state,
        id: '',
        name: '',
        email: ''
      }

    default:
      return state
  }
}

export const initialState: User = {
  // id: localStorage.getItem('id') as string,
  id: document.cookie.split(';').find((x: String) => x.match("id="))?.split("=")[1],
  name: localStorage.getItem('name') as string,
  email: localStorage.getItem('email') as string
}