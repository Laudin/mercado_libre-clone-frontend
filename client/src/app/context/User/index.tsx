import React from 'react';
import { reducer, initialState } from './reducer'

export const UserContext = React.createContext({
  currentUser: initialState,
  dispatch: reducer
})

export const UserProvider = ({ children }) => {
  
  const [currentUser, dispatch] = React.useReducer(reducer, initialState);
  
  return (
    <UserContext.Provider value={{currentUser, dispatch}}>
      {children}
    </UserContext.Provider>
  )
}