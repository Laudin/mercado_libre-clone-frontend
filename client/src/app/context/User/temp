export const reducer = (state, action: {type: string}) => {
  console.log(action)
  switch (action.type) {
    case "set_user":
      console.log('login')
      localStorage.setItem('name', state.name);
      localStorage.setItem('email', state.email);
      return {
        ...state,
        name: state.name,
        email: state.email
      }
    case "clear_user":
      console.log('logout')
      localStorage.removeItem('name');
      localStorage.removeItem('email');
      return {
        ...state,
        name: '',
        email: ''
      }
    default:
      return state
  }
}

export interface user {
  name: string,
  email: string
}

export const initialState: any = {
  name: '',
  email: ''
}

///////////////////////////////////

import React from 'react';
import { reducer, initialState, user } from './reducer'

export const UserContext = React.createContext({
  currentUser: initialState,
  dispatch: reducer
})

export const UserProvider = ({ children }) => {
  
  const [_, dispatch] = React.useReducer(reducer, initialState);

  const currentUser: user | any = () => {
    const name = localStorage.getItem('name') as string;
    const email = localStorage.getItem('email') as string;
    return {
      name: name,
      email: email
    };
  };
  /* console.log('currentUser: ')
  console.log(currentUser) */
  const setCurrentUser = () => {
    
  }
  
  return (
    <UserContext.Provider value={{currentUser, dispatch}}>
      {children}
    </UserContext.Provider>
  )
  /* const getUser = () => {
    const name = localStorage.getItem('name');
    const email = localStorage.getItem('email');
    //const userToken = JSON.stringify(tokenString as string); //pff alpedo
    //console.log(tokenString);
    return {
      name: name,
      email: email
    };
  };
   const saveUser = (name, email) => {
    localStorage.setItem('name', name);
    localStorage.setItem('email', email);
    setCurrentUser({
      name: name,
      email: email
    });
  }; */



  /* return {
    setCurrentUser: saveUser,
    currentUser,
  }; */
}