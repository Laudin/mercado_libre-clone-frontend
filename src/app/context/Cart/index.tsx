import React from 'react';
import { reducer, initialState } from './reducer'
import * as api from '../../api/usersApi'

export const CartContext = React.createContext({
   cart: initialState,
   dispatchCart: reducer
})

export const CartProvider = ({ children }) => {

   const [cart, dispatchCart] = React.useReducer(reducer, initialState);

   // infinite loop..
   // api.getCart().then(res => dispatchCart({ type: "initialize", product: '', initialValue: res }))

   return (
      <CartContext.Provider value={{ cart, dispatchCart }}>
         {children}
      </CartContext.Provider>
   )
}