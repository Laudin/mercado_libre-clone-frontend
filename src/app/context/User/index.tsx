import React from 'react';
import { reducer, initialState } from './reducer'

export const UserContext = React.createContext({
  currentUser: initialState,
  dispatchUser: reducer
})

export const UserProvider = ({ children }) => {

  const [currentUser, dispatchUser] = React.useReducer(reducer, initialState);

  return (
    <UserContext.Provider value={{ currentUser, dispatchUser }}>
      {children}
    </UserContext.Provider>
  )
}