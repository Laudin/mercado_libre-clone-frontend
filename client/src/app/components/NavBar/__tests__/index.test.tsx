import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { fireEvent, render, screen } from '@testing-library/react'
import 'jest-styled-components'

import { BrowserRouter } from 'react-router-dom';
import { UserProvider } from '../../../context/User'

import { NavBar } from '../index';

const shallowRenderer = createRenderer()

describe('<NavBar />', () => {
   test('should render', () => {
      shallowRenderer.render(
         <BrowserRouter>
            <UserProvider>
               <NavBar />
            </UserProvider>
         </BrowserRouter>
      );
      const rendererOutput = shallowRenderer.getRenderOutput();
      expect(rendererOutput).toMatchSnapshot()
   })

   //this works
   it('should show input with initial value set', async () => {
      render(<input type="text" data-testid="form-field-firstname" />);
      const inputField = screen.getByTestId("form-field-firstname")
      fireEvent.change(inputField, { target: { value: 'a' } })
      expect(inputField).toHaveDisplayValue('a');
   });
   /*
   //with redux
   it('should dispatch action on username change', () => {
   const input = component.container.querySelector('input');
   fireEvent.change(input!, { target: { value: 'test' } });
   expect(store.getState().githubRepoForm.loading).toBe(true);
 }); */
});