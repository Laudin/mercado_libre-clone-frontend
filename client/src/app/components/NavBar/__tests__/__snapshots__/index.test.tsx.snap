// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> should match snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 auto;
  margin-top: 5px;
  width: 95%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 40px;
}

.c4 {
  -webkit-flex: 0.5;
  -ms-flex: 0.5;
  flex: 0.5;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  gap: 10px;
  list-style: none;
}

.c17 {
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  margin: 0;
}

.c9 {
  margin: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 auto;
  margin-bottom: 5px;
  width: 95%;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  gap: 10px;
}

.c15 {
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c13 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: lightgray;
  border-radius: 5px;
  width: 200px;
  padding: 0px;
  visibility: hidden;
  left: -50%;
  z-index: 99;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c13 .c14 {
  padding: 10px;
  border-radius: 5px;
}

.c13 .c14:hover {
  background: lightskyblue;
}

.c11 {
  position: relative;
}

.c11:hover .c12 {
  visibility: visible;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff159;
  padding: 0 20px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <img
        className="c2"
        src="logo__large_plus.png"
      />
    </a>
    <div
      className="c3"
    >
      <img
        className=""
        src="#"
      />
      <input
        className="c4"
        placeholder="Buscar.."
      />
    </div>
    <p>
      Envíos gratis en 24 hs a partir de $4.000
    </p>
  </div>
  <div
    className="c5"
  >
    <div
      className="c6"
    >
      <img
        className=""
        src=""
      />
      <div
        className="c7"
      >
        <p
          className="c8"
        >
          Enviar a
        </p>
        <p
          className="c9"
        >
          Capital Federal
        </p>
      </div>
    </div>
    <div
      className="c10"
    >
      <div
        className="c11"
      >
        Categorias
        <ul
          className="c12 c13"
        >
          <a
            className="c14 c15"
            href="/Vehículos"
            onClick={[Function]}
          >
            Vehículos
          </a>
          <a
            className="c14 c15"
            href="/Inmuebles"
            onClick={[Function]}
          >
            Inmuebles
          </a>
          <a
            className="c14 c15"
            href="/Supermercado"
            onClick={[Function]}
          >
            Supermercado
          </a>
          <a
            className="c14 c15"
            href="/Farmacia"
            onClick={[Function]}
          >
            Farmacia
          </a>
          <a
            className="c14 c15"
            href="/Tecnología"
            onClick={[Function]}
          >
            Tecnología
          </a>
          <a
            className="c14 c15"
            href="/Herramientas"
            onClick={[Function]}
          >
            Herramientas
          </a>
          <a
            className="c14 c15"
            href="/Construcción"
            onClick={[Function]}
          >
            Construcción
          </a>
          <a
            className="c14 c15"
            href="/Deportes y Fitness"
            onClick={[Function]}
          >
            Deportes y Fitness
          </a>
          <a
            className="c14 c15"
            href="/Moda"
            onClick={[Function]}
          >
            Moda
          </a>
          <a
            className="c14 c15"
            href="/Juegos y Juguetes"
            onClick={[Function]}
          >
            Juegos y Juguetes
          </a>
          <a
            className="c14 c15"
            href="/Belleza y Ciudado Personal"
            onClick={[Function]}
          >
            Belleza y Ciudado Personal
          </a>
        </ul>
      </div>
      <a
        className="c14 c15"
        href="/"
        onClick={[Function]}
      >
        Ofertas
      </a>
      <a
        className="c14 c15"
        href="/"
        onClick={[Function]}
      >
        Historial
      </a>
      <a
        className="c14 c15"
        href="/supermercado"
        onClick={[Function]}
      >
        Supermercado
      </a>
      <a
        className="c14 c15"
        href="/Moda"
        onClick={[Function]}
      >
        Moda
      </a>
      <a
        className="c14 c15"
        href="/publish"
        onClick={[Function]}
      >
        Vender
      </a>
      <a
        className="c14 c15"
        href="/"
        onClick={[Function]}
      >
        Ayuda
      </a>
    </div>
    <div
      className="c16"
    >
      <a
        className="c17"
        onClick={[Function]}
      >
        Crear cuenta
      </a>
      <a
        className="c17"
        onClick={[Function]}
      >
        Ingresar
      </a>
      <a
        className="c17"
      >
        Mis Compras
      </a>
      <a
        className="c17"
      >
        Carrito
      </a>
    </div>
  </div>
</div>
`;

exports[`<NavBar /> should render 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <UserProvider>
    <NavBar />
  </UserProvider>
</Router>
`;
